### PR TITLE
Add `linalg.multi_dot` API.

### PR DESCRIPTION
**Pull request by khushi-411**
Hi, Team!
The PR aims to add `cupy.linalg.multi_dot` API. It follows #6078.
I would love to learn from your viewpoints.

~TODO: Add tests.~

Thanks!

### CODE REVIEW

File: cupy/linalg/_product.py [hunk 1]
```diff
@@ -336,51 +401,113 @@ def tensordot(a, b, axes=2):
...
+def multi_dot(arr, *, out=None):
+    """Compute the dot product of two or more arrays in a single function
+    call, and automatically select the fastest evaluation order.
+
+    Note: `multi_dot` chains `cupy.dot` and uses optimal parenthesization
+    of the matrices. Depending on the shape of the matrices, this can
+    speed up the multiplication a lot.
+
+    Parameters
+    ----------
```
Comment (emcastillo) [inline cupy/linalg/_product.py:457]: Please keep docstring format consistent with the codebase 🙂

File: cupy/linalg/_product.py [hunk 2]
```diff
@@ -42,6 +42,71 @@
...
+    # To store dimensions of the matrices
+    p = [a.shape[0] for a in arr] + [arr[-1].shape[1]]
+    # Matrix of costs of the subproblems
+    # m[i,j]: min number of scalar multiplications needed to compute A_{i..j}
+    m = cupy.zeros((n, n), dtype=cupy.double)
+    # Actual ordering
+    # s[i, j] is the value of k at which we split the product A_i..A_j
+    s = cupy.empty((n, n), dtype=cupy.intp)
+
+    for l in range(1, n):
```
Comment (emcastillo) [inline cupy/linalg/_product.py:84]: if `arr` is 2d I feel like this code is going to be really slow
Do you have performance measures of the routine compared to NumPy?

File: cupy/linalg/_product.py [hunk 3]
```diff
@@ -42,6 +42,71 @@
...
+        for i in range(n - l):
+            j = i + l
+            m[i, j] = cupy.Inf
+            for k in range(i, j):
+                q = m[i, k] + m[k+1, j] + p[i]*p[k+1]*p[j+1]
+                if q < m[i, j]:
```
Comment (emcastillo) [inline cupy/linalg/_product.py:90]: this imposes device/cpu sync.
I think this function should be rewritten into a Cuda kernel using `RawKernel`

File: cupy/linalg/_product.py (hunk 2 above)
Comment (khushi-411) [reply cupy/linalg/_product.py:84]: I'll check that and update here soon :)

**Comment by emcastillo** [comment]
Closing, feel free to reopen

**Pull request closed without merging**